WebServices/AzureSignalR/ChatClient/MainPage.xaml.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using Xamarin.Forms;
+
+namespace ChatClient
+{
+    public partial class MainPage : ContentPage
+    {
+        SignalRService signalR;
+
+        public MainPage()
+        {
+            InitializeComponent();
+
+            signalR = new SignalRService();
+            signalR.Connected += SignalR_ConnectionChanged;
+            signalR.ConnectionFailed += SignalR_ConnectionChanged;
+            signalR.NewMessageReceived += SignalR_NewMessageReceived;
+        }
+
+        private void AddMessage(string message)
+        {
+            Device.BeginInvokeOnMainThread(() =>
+           {
+               Label label = new Label
+               {
+                   Text = message,
+                   HorizontalOptions = LayoutOptions.Start,
+                   VerticalOptions = LayoutOptions.Start
+               };
+
+               messageList.Children.Add(label);
+           });
+        }
+
+        private void SignalR_NewMessageReceived(object sender, Model.Message message)
+        {
+            string msg = $"{message.Name} ({message.TimeReceived}) - {message.Text}";
+            AddMessage(msg);
+        }
+
+        private void SignalR_ConnectionChanged(object sender, bool success, string message)
+        {
+            Device.BeginInvokeOnMainThread(() =>
```

Comment:
Again, just double checking that marshalling is required.

Reply:
Yep! I actually ran into this one while testing that events fire correctly. This handler makes UI changes and throws Exceptions when called off-thread.